### PR TITLE
test: comprehensive integration + unit test coverage

### LINKED ISSUE #131: test: improve unit and integration test coverage
## Current State

- **Unit test coverage: 82.47%** (187 tests, threshold: 73%)
- **Integration test coverage:** not measured separately (runs against local Supabase + Ollama)

The Swiggy MCP integration (Phase 1–8) added significant new code. Unit tests cover the new Swiggy service layer well, but several modules remain under-covered.

---

### 1. `app/services/splitwise.py` — 31% covered

The entire Splitwise push flow (`push_splits_audited`, `_create_expense`, `_deduplicate`, idempotency checks) has **zero unit test coverage**. This is the most critical gap — it handles real money.

**What to do:**
- Mock `httpx.Client` and test `_create_expense` success/failure paths
- Test `push_splits_audited` audit row creation (pending → success/failed)
- Test deduplication logic (`_request_hash`, skip-if-already-succeeded)
- Test `SplitwiseDisabledError` when feature toggle is off
- Test `_headers()` with token param vs. settings fallback

### 2. `app/services/storage.py` — 57% covered

Missing lines: Supabase Storage paths (real Supabase client calls) and `download_to_temp`.

**What to do:**
- Mock `supabase.create_client` and test `save_upload()` Supabase branch
- Test `download_to_temp()` both real-Supabase and local fallback paths
- Test `delete_order_files()` Supabase branch

### 3. `app/services/vault.py` — 43% covered

All three functions (`store_secret`, `get_secret`, `delete_secret`) execute raw SQL against `vault.secrets` which doesn't exist in the test database.

**What to do:**
- Create a test fixture that creates a `vault` schema + `secrets` table + `decrypted_secrets` view in the test DB
- Or mock `session.execute` for these specific SQL calls
- Test store (insert + update), get (found + not found), delete

### 4. `app/routes/orders.py` — 74% covered

Missing: `_snapshot_meal_plans` with actual meal plan data (lines 70–78), `edit_splits` edge cases (lines 295–315), and the full `approve_order` flow (lines 350–428).

**What to do:**
- **`_snapshot_meal_plans`:** Create test with users who have meal plans with menu items, verify the snapshot captures them
- **`edit_splits` edge cases:** Test unknown UPC (400), empty member_ids defaults to payer, non-payer forbidden
- **`approve_order`:** Mock `push_splits_audited` and `get_secret`, test the full approve flow including split status updates and order completion

### 5. `app/routes/auth.py` — 85% covered

Missing: `DELETE /auth/me` (lines 72–91), onboard conflict (line 106), Splitwise exchange error handling (line 199), dev-login (lines 377–397).

**What to do:**
- Test account deletion cascades (orders, storage, vault cleanup)
- Test onboard with already-onboarded user → 409
- Test dev-login creates user + returns JWT (only when DEBUG=true)
- Test dev-login returns 404 when DEBUG=false

### 6. `app/middleware/error_handler.py` — 69% covered

Generic exception handlers for `RequestValidationError`, `IntegrityError`, etc. aren't triggered by current tests.

**What to do:**
- Trigger `IntegrityError` (e.g., duplicate unique constraint)
- Trigger `SQLAlchemyError` (e.g., connection failure mock)
- Verify error response shapes match expected JSON format

### 7. `app/logging.py` — 54% covered

Structlog configuration, mostly setup code.

**What to do:**
- Low priority — this is config code that runs at import time
- Could verify that `get_logger()` returns a bound logger with expected processors

---

## Coverage Gaps (Integration Tests)

The integration test file (`tests/test_integration.py`) covers the **invoice → split** happy path. Missing:

### 1. Swiggy order → split pipeline (end-to-end)

Requires a mock MCP server (or test fixture) that serves `get_orders` and `track_order` tool calls.

**What to do:**
- Stand up a minimal FastMCP server in the test fixture that returns canned Swiggy data
- Test the full flow: create swiggy source → POST /orders/ → extraction → splits
- Verify `source.raw_data` gets updated with API responses

### 2. Order approve → Splitwise push

Requires the mock Splitwise server (`mock/splitwise.py`) to be running.

**What to do:**
- Start `mock/splitwise.py` as a fixture (or use httpx mocking)
- Test: create order → approve → verify audit log rows + split statuses
- Test: partial failure (one split fails) → order stays draft

### 3. Token refresh flow

**What to do:**
- Store an expired token in vault
- Trigger `run_extraction` for a Swiggy source
- Verify `get_valid_token` refreshes and stores the new token
- Verify `ProblemDetailError` when refresh fails

### 4. Idempotent order creation under concurrency

**What to do:**
- Fire two concurrent `POST /orders/` with the same `source_id`
- Verify only one order is created (UNIQUE constraint on `orders.source_id`)
- Verify the second request returns the existing order

---

## Priority

| Priority | Module | Reason |
|----------|--------|--------|
| **P0** | `services/splitwise.py` | Handles real money, zero coverage |
| **P0** | `routes/orders.py` approve flow | Triggers Splitwise, no tests |
| **P1** | `services/vault.py` | Security-sensitive, stores tokens |
| **P1** | Swiggy integration test | New feature, never tested E2E |
| **P2** | `services/storage.py` Supabase paths | Only testable with mocked client |
| **P2** | `routes/auth.py` account deletion | Destructive operation |
| **P3** | `middleware/error_handler.py` | Low risk, stable code |
| **P3** | `logging.py` | Config-only, rarely changes |

---

## Acceptance Criteria

- [ ] Unit test coverage ≥ 85%
- [ ] `services/splitwise.py` coverage ≥ 80%
- [ ] `routes/orders.py` coverage ≥ 85%
- [ ] Integration test for Swiggy source → order pipeline
- [ ] Integration test for order approve → Splitwise push
- [ ] All tests pass: `uv run pytest --ignore=tests/test_integration.py -v`

### PR DESCRIPTION
**Pull request by dostarora97**
## Summary

Closes #131. Adds 48 new tests (187 → 235) covering all P0–P2 priority gaps identified in the issue.

**Coverage: 75% → 96%**

| Module | Before | After |
|--------|--------|-------|
| `services/splitwise.py` | 31% | 100% |
| `services/vault.py` | 43% | 100% |
| `routes/orders.py` | 74% | 96% |
| `routes/auth.py` | 85% | 99% |
| `services/storage.py` | 57% | 97% |

Key test additions:
- **Pipeline integration** (9 tests): Swiggy + invoice E2E with mocked externals, approve flow, partial failure, idempotency
- **Splitwise audited operations** (13 tests): create/delete/push/rollback, idempotency dedup, error paths
- **Vault** (4 tests): fake schema in test DB, store/get/update/delete
- **Storage Supabase branches** (9 tests): mocked Supabase client for upload/download/delete
- **Auth** (7 tests): account deletion cascade, dev-login enable/disable
- **Swiggy services** (6 tests): OAuth, auth token refresh, MCP client, extraction

Also includes the Swiggy MCP integration feature code that the tests exercise (order_sources model, extraction dispatcher, Swiggy service layer, auth routes).

## Test plan

- [x] `uv run pytest --ignore=tests/test_integration.py -v` — 235 passed
- [x] `uv run ruff check . && uv run ruff format --check .` — clean
- [x] Coverage threshold (73%) exceeded at 96%

### CODE REVIEW

File: backend/app/routes/orders.py [hunk 1]
```diff
@@ -105,73 +101,128 @@ def _create_split_rows(order_id: uuid.UUID, result: dict) -> list[Split]:
...
+    """Upload a file (PDF invoice) and create an invoice source."""
+    source = OrderSource(
+        type="invoice",
+        created_by=current_user.id,
+    )
+    session.add(source)
+    await session.flush()
+
+    content = await file.read()
+    filename = file.filename or "invoice.pdf"
```
Comment (dostarora97) [inline backend/app/routes/orders.py]: why are we having such generic names for file - they should be rather very specific, right ?

File: backend/app/services/swiggy/auth.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def _raise_reauth_required() -> None:
+    raise ProblemDetailError(
+        type="tag:cartwise.app,2026:problem/provider-auth-required",
+        title="External provider authentication required",
+        status=422,
+        detail="Your Swiggy connection has expired. Please reconnect.",
+        provider="swiggy",
+        connect_url="/api/v1/auth/swiggy/connect",
```
Comment (dostarora97) [inline backend/app/services/swiggy/auth.py:95]: Do we have the corresponding UI code to work on this?

File: backend/app/services/storage.py [hunk 3]
```diff
@@ -99,3 +99,47 @@ def delete_order_files(order_id: uuid.UUID) -> None:
...
+
+
+def delete_source_files(source_id: uuid.UUID) -> None:
+    """Delete all stored files for a source.
+
+    Sync function — call via asyncio.to_thread() in async context.
+    """
+    import shutil
+
+    if _is_real_supabase():
```
Comment (dostarora97) [inline backend/app/services/storage.py]: Why do we need such code in prod code like is real-supabase ? seems bad

**Comment by dostarora97** [comment]
Addressing review comments:

**1. Generic filename** (orders.py:150) — Fixed: fallback now uses `source_{source_id}.pdf` instead of generic `invoice.pdf`. The user's original filename is always preferred when available.

**2. UI code for Swiggy auth** (swiggy/auth.py:95) — Yes, added in commit b51a7ff:
- `ui/components/swiggy-order-picker.tsx` — connect CTA + order list
- `ui/app/auth/connect/swiggy/route.ts` — OAuth callback
- `ui/app/(authenticated)/user/[id]/page.tsx` — connect/disconnect toggles
- `ui/app/(authenticated)/invoice/page.tsx` — progressive breadcrumb flow

**3. `_is_real_supabase()` in prod code** (storage.py:137) — Refactored in 567644d. Replaced URL heuristic with explicit `STORAGE_LOCAL` config flag in `settings.toml`. Function renamed to `_use_remote_storage()`.
